File: 05-DIP/python/refactored.py
from abc import ABC, abstractmethod

class PaymentGateway(ABC):
    @abstractmethod
    def charge(self, amount):
        pass

class RazorpayGateway(PaymentGateway):
    def charge(self, amount):
        print(f"Charging ₹{amount} using Razorpay")

class GooglepayGateway(PaymentGateway):
    def charge(self, amount):
        print(f"Charging ₹{amount} using Googlepay")

class PaypalGateway(PaymentGateway):
    def charge(self, amount):
        print(f"Charging ₹{amount} using Paypal")

class PaymentService:
    def __init__(self, gateway):
        self.gateway = gateway

    def pay(self, amount):
        self.gateway.charge(amount)

if __name__ == "__main__":
    razorpay = RazorpayGateway()
    service = PaymentService(razorpay)
    service.pay(1000)

    gpay = GooglepayGateway()
    service = PaymentService(gpay)
    service.pay(2000)

'''
PS C:\Users\Shikha Pandey\Documents\github\solid-design-principles>python 05-DIP/python/refactored.py"
Charging ₹1000 using Razorpay
Charging ₹2000 using Googlepay
'''

'''
Notice the important change:

# Before
self.gateway = RazorpayGateway()

# After
self.gateway = gateway

PaymentService no longer creates or knows about the concrete gateway.
'''

File: 05-DIP/cpp/refactored.cpp
#include <iostream>

class PaymentGateway {
public:
    virtual void charge(double amount) = 0;
    virtual ~PaymentGateway() = default;
};

class RazorpayGateway : public PaymentGateway {
public:
    void charge(double amount) override {
        std::cout << "Charging " << amount
                  << " using Razorpay\n";
    }
};

class GPayGateway : public PaymentGateway {
public:
    void charge(double amount) override {
        std::cout << "Charging " << amount
                  << " using GPay\n";
    }
};

class PayPalGateway : public PaymentGateway {
public:
    void charge(double amount) override {
        std::cout << "Charging " << amount
                  << " using PayPal\n";
    }
};

class PaymentService {
private:
    PaymentGateway& gateway;

public:
    PaymentService(PaymentGateway& gateway)
        : gateway(gateway) {}

    void pay(double amount) {
        gateway.charge(amount);
    }
};

int main() {
    RazorpayGateway razorpay;
    PaymentService service(razorpay);
    service.pay(1000);

    GPayGateway gpay;
    PaymentService gpayService(gpay);
    gpayService.pay(2000);

    return 0;
}